sys/auto_init/saul/auto_init_io1_xplained.c

```diff
@@ -73,8 +73,8 @@ void auto_init_io1_xplained(void)
 
         /* GPIOs */
         for (unsigned j = 1; j < 4; j++) {
-            saul_entries[i * 4 + j].dev = &(io1_xplained_saul_gpios[j]);
-            saul_entries[i * 4 + j].name = io1_xplained_saul_info[i][j].name;
+            saul_entries[i * 4 + j].dev = &(io1_xplained_saul_gpios[j - 1]);
+            saul_entries[i * 4 + j].name = io1_xplained_saul_info[i][j - 1].name;
```

Comment:
I know this is maybe a hair splitting style thing, but I think this off-by-one error could have been easier to spot if the loop counter would start at 0 (and it would IMHO also the code easier to follow, as a C programmer usually expect that). I.e. I would restructure the whole loop like this:

```C
        for (unsigned j = 0;
             j < (sizeof(io1_xplained_saul_gpio)/sizeof(io1_xplained_saul_gpio[0]));
             j++) {
            saul_reg_t *saul_entry = saul_entries[i * 4 + j + 1];

            saul_entry->dev = &(io1_xplained_saul_gpios[j]);
            saul_entry->name = io1_xplained_saul_info[i][j].name;
            saul_entry->driver = &gpio_out_saul_driver;
            saul_reg_add(saul_entry));
        }
```

Reply:
BTW, I think you explained this once to me, but I already forgot: What does the 4 stand for? Maybe make this clearer or a macro. Also in https://github.com/aabadie/RIOT/blob/4c36321ca07f5d1f1610d0a436d714ef173f3a0d/sys/auto_init/saul/auto_init_io1_xplained.c#L41.

Reply:
4 stands for the number of devices (temperature sensor, gpios) exposed to SAUL by the driver.

Reply:
> I know this is maybe a hair splitting style thing

I wanted the loop to start at 1 to keep some kind of continuity of the registered devices in the saul registry: 0 is for the temperature sensor, 1 for GPIO LED, 2 and 3 for GPIO pins

But ok, your suggestion makes sense.

Reply:
> 4 stands for the number of devices (temperature sensor, gpios) exposed to SAUL by the driver.

Then why don't you use `IO1_XPLAINED_NUM` here (which according to doc this is exactly that)? Why is `IO1_XPLAINED_NUM` again multiplied by 4 in the definition `saul_entries`?

Reply:
`IO1_XPLAINED_NUM` is number of extension boards (device) with 4 sub-devices (temperature and gpios). So no, we can't use this.